### PR TITLE
PorousFlow: hysteretic relative permeability materials added 

### LINKED ISSUE #16324: PorousFlow: add Materials to compute hysteretic relative permeability
## Reason

Final stage in the hysteresis work: need to have a user interface (Material) to compute the relative permeabilities

## Design

Something like `PorousFlowHystereticRelativePermeabilityLiquid` and `PorousFlowHystereticRelativePermeabilityGas`.

## Impact

Ability to use hysteretic relative permeabilities

### PR DESCRIPTION
**Pull request by WilkAndy**
Fixes #16324

### CODE REVIEW

**Comment by WilkAndy** [comment]
@cpgr - this is the final installment of the hysteresis stuff.  The code is pretty straightforward.  Do you think you could have a look?

File: modules/porous_flow/doc/content/source/materials/PorousFlowHystereticRelativePermeabilityGas.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# PorousFlowHystereticRelativePermeabilityGas
+
+This Material computes the gas relative permeability using the van-Genuchten formulation.  The relative permeability is assumed to be hysteretic.  This Material may be used two-phase situations only.
```
Comment (cpgr) [inline modules/porous_flow/doc/content/source/materials/PorousFlowHystereticRelativePermeabilityGas.md:3]:
```suggestion
This Material computes the gas relative permeability using the van Genuchten formulation.  The relative permeability is assumed to be hysteretic.  This Material may be used in two-phase situations only.
```

File: modules/porous_flow/doc/content/source/materials/PorousFlowHystereticRelativePermeabilityLiquid.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# PorousFlowHystereticRelativePermeabilityLiquid
+
+This Material computes the liquid relative permeability using the van-Genuchten formulation.  The relative permeability is assumed to be hysteretic.  This Material may be used in single-phase or two-phase situations only.
```
Comment (cpgr) [inline modules/porous_flow/doc/content/source/materials/PorousFlowHystereticRelativePermeabilityLiquid.md:3]:
```suggestion
This Material computes the liquid relative permeability using the van Genuchten formulation.  The relative permeability is assumed to be hysteretic.  This Material may be used in single-phase or two-phase situations only.
```

File: modules/porous_flow/include/materials/PorousFlowHystereticRelativePermeabilityLiquid.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  : public PorousFlowHystereticRelativePermeabilityBase
+{
+public:
+  static InputParameters validParams();
+
+  PorousFlowHystereticRelativePermeabilityLiquid(const InputParameters & parameters);
+
+protected:
+  /// Phase number of liquid phase
+  const unsigned _liquid_phase;
```
Comment (cpgr) [inline modules/porous_flow/include/materials/PorousFlowHystereticRelativePermeabilityLiquid.h:27]: Could you put `_liquid_phase` into the base class here and remove it from the gas and liquid classes?

File: modules/porous_flow/src/materials/PorousFlowHystereticRelativePermeabilityGas.C [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      "PorousFlow material that computes relative permeability of the gas phase in 1-phase or "
+      "2-phase models that include hysteresis.  You should ensure that the 'phase' for this "
+      "Material does indeed represent the gas phase");
+  return params;
+}
+
+PorousFlowHystereticRelativePermeabilityGas::PorousFlowHystereticRelativePermeabilityGas(
+    const InputParameters & parameters)
+  : PorousFlowHystereticRelativePermeabilityBase(parameters),
+    _liquid_phase(_phase_num == 0 ? 1 : 0),
```
Comment (cpgr) [inline modules/porous_flow/src/materials/PorousFlowHystereticRelativePermeabilityGas.C:46]: Maybe this prevents you from putting `_liquid_phase` into the base class? If so, ignore that comment

**Review by cpgr — DISMISSED** [review]

File: modules/porous_flow/doc/content/source/materials/PorousFlowHystereticRelativePermeabilityGas.md (hunk 1 above)
Comment (WilkAndy) [reply modules/porous_flow/doc/content/source/materials/PorousFlowHystereticRelativePermeabilityGas.md:3]: What is it with me a van-HYPHEN-Genuchten ?!?!

File: modules/porous_flow/doc/content/source/materials/PorousFlowHystereticRelativePermeabilityLiquid.md (hunk 2 above)
Comment (WilkAndy) [reply modules/porous_flow/doc/content/source/materials/PorousFlowHystereticRelativePermeabilityLiquid.md:3]: done

File: modules/porous_flow/src/materials/PorousFlowHystereticRelativePermeabilityGas.C (hunk 4 above)
Comment (WilkAndy) [reply modules/porous_flow/src/materials/PorousFlowHystereticRelativePermeabilityGas.C:46]: yea, that's the reason i can't put `_liquid_phase` in the base class.

File: modules/porous_flow/include/materials/PorousFlowHystereticRelativePermeabilityLiquid.h (hunk 3 above)
Comment (WilkAndy) [reply modules/porous_flow/include/materials/PorousFlowHystereticRelativePermeabilityLiquid.h:27]: bummer that i can't do this without being rather obscure (in my opinion)

**Comment by WilkAndy** [comment]
Hi @aeslaughter , i think this is ready to be merged.

File: modules/porous_flow/doc/content/source/materials/PorousFlowHystereticRelativePermeabilityGas.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+This Material computes the gas relative permeability using the van Genuchten formulation.  The relative permeability is assumed to be hysteretic.  This Material may be used in two-phase situations only.
+
+Along the drying curve, the relative permeability is
+\begin{equation}
+k_{r, g} = k_{r, g}^{max}\left(1 - \bar{S}_{l}\right)^{\gamma} \left( 1- \bar{S}_{l}^{1/m} \right)^{2m}
+\end{equation}
+where
+\begin{equation}
+\bar{S}_{l} = \frac{S_{l} - S_{l, r}}{1 - S_{l, r}}
+\end{equation}
```
Comment (aeslaughter) [inline modules/porous_flow/doc/content/source/materials/PorousFlowHystereticRelativePermeabilityGas.md:12]: FYI, we are moving away from this syntax. Please use `!equation` in the future.

```
!equation
k_{r, g} = k_{r, g}^{max}\left(1 - \bar{S}_{l}\right)^{\gamma} \left( 1- \bar{S}_{l}^{1/m} \right)^{2m}
```

**Review by aeslaughter — APPROVED** [review]

**Pull request merged**